src/TypeAdapter/ArrayTypeAdapter.php

```diff
@@ -0,0 +1,210 @@
+
+    /**
+     * @var TypeToken
+     */
+    protected $keyType;
+
+    /**
+     * @var TypeAdapter
+     */
+    protected $valueTypeAdapter;
+
+    /**
+     * @var int
+     */
+    protected $numberOfGenerics;
+
+    /**
+     * A TypeAdapter cache keyed by raw type
+     *
+     * @var TypeAdapter[]
+     */
+    protected $adapters = [];
+
+    /**
+     * Constructor
+     *
+     * @param TypeAdapterProvider $typeAdapterProvider
+     * @param TypeToken $keyType
+     * @param TypeAdapter $valueTypeAdapter
+     * @param int $numberOfGenerics
+     */
+    public function __construct(
+        TypeAdapterProvider $typeAdapterProvider,
+        TypeToken $keyType,
+        TypeAdapter $valueTypeAdapter,
+        int $numberOfGenerics
+    ) {
+        $this->typeAdapterProvider = $typeAdapterProvider;
+        $this->keyType = $keyType;
+        $this->valueTypeAdapter = $valueTypeAdapter;
+        $this->numberOfGenerics = $numberOfGenerics;
+    }
+
+    /**
+     * Read the next value, convert it to its type and return it
+     *
+     * @param array|null $value
+     * @param ReaderContext $context
+     * @return array|null
+     */
+    public function read($value, ReaderContext $context): ?array
+    {
+        if ($value === null) {
+            return null;
+        }
+
+        if (!is_array($value)) {
+            throw new JsonSyntaxException(sprintf('Could not parse json, expected array or object but found "%s"', gettype($value)));
+        }
+
+        $result = [];
+
+        if ($this->numberOfGenerics > 2) {
+            throw new LogicException('Array may not have more than 2 generic types');
+        }
+
+        if ($this->keyType->phpType !== TypeToken::WILDCARD
+            && $this->keyType->phpType !== TypeToken::STRING
+            && $this->keyType->phpType !== TypeToken::INTEGER
+        ) {
+            throw new LogicException('Array keys must be strings or integers');
+        }
+
+        $arrayIsObject = $this->numberOfGenerics === 2 || is_string(key($value));
+        $enableScalarAdapters = $context->enableScalarAdapters();
+
+        foreach ($value as $key => $item) {
+            $itemValue = null;
+            switch ($this->numberOfGenerics) {
+                case 0:
+                    if (!$enableScalarAdapters && is_scalar($item)) {
+                        $itemValue = $item;
+                        break;
+                    }
+
+                    if (!$arrayIsObject) {
+                        $itemValue = $this->valueTypeAdapter->read($item, $context);
+                        break;
+                    }
+
+                    if (is_array($item)) {
+                        $itemValue = $this->read($item, $context);
+                        break;
+                    }
+
+                    $type = TypeToken::createFromVariable($item);
+                    $adapter = $this->adapters[$type->rawType] ?? $this->adapters[$type->rawType] = $this->typeAdapterProvider->getAdapter($type);
```

Comment:
I think this could be it's own function within this class as it is used a few times and would make it easier to read.

something just like `getRawTypeAdapter`?